README.md
[Oath Container Library](http://www.aolplatforms.com/header-bidding-solution)
===========================

[![Build Status](https://api.travis-ci.org/aol/Prebid.js.svg?branch=master)](https://travis-ci.org/aol/Prebid.js)

This is a fork of Prebid.js made by [AOL](http://www.aolplatforms.com/).
Check the [AOL Platforms website](http://www.aolplatforms.com/header-bidding-solution) for documentation
and more information on its unique features. Here is just a brief list of differences to the official Prebid:

- AOL adapter
    - `onemobile` - alias for One Mobile endpoint
    - `onedisplay` - alias for One Display endpoint
- Multiple versions available, CDN hosted
- Auto-updates for
    - New backwards-compatible features
    - Bug fixes
- Dynamic bidder injection, enabling an optimal download size
- Reports on how the auctions are performing
- Version numbers separate from official Prebid

Container library is available on the following URL (works fine with http as well as https):

    vlibs.advertising.com/prebid/adapters=appnexus,openx;/prebid-2.x.x.js

This URL is designed specifically for the Container library and it supports two magic features:

1. Request time injection of Prebid adapters (AOL adapter included by default)
2. Delivery of latest available major version (2.x.x)

**Here continues the original README:**

---


[![Build Status](https://api.travis-ci.org/prebid/Prebid.js.svg?branch=master)](https://travis-ci.org/prebid/Prebid.js)
[![Percentage of issues still open](http://isitmaintained.com/badge/open/prebid/Prebid.js.svg)](http://isitmaintained.com/project/prebid/Prebid.js "Percentage of issues still open")
[![Average time to resolve an issue](http://isitmaintained.com/badge/resolution/prebid/Prebid.js.svg)](http://isitmaintained.com/project/prebid/Prebid.js "Average time to resolve an issue")
[![Code Climate](https://codeclimate.com/github/prebid/Prebid.js/badges/gpa.svg)](https://codeclimate.com/github/prebid/Prebid.js)
[![Coverage Status](https://coveralls.io/repos/github/prebid/Prebid.js/badge.svg)](https://coveralls.io/github/prebid/Prebid.js)
[![devDependencies Status](https://david-dm.org/prebid/Prebid.js/dev-status.svg)](https://david-dm.org/prebid/Prebid.js?type=dev)

# Prebid.js

> A free and open source library for publishers to quickly implement header bidding.

This README is for developers who want to contribute to Prebid.js.
Additional documentation can be found at [the Prebid homepage](http://prebid.org).
Working examples can be found in [the developer docs](http://prebid.org/dev-docs/getting-started.html).

**Table of Contents**

- [Install](#Install)
- [Build](#Build)
- [Run](#Run)
- [Contribute](#Contribute)

<a name="Install"></a>

## Install

    $ git clone https://github.com/prebid/Prebid.js.git
    $ cd Prebid.js
    $ npm install

*Note:* You need to have `NodeJS` 4.x or greater installed.

<a name="Build"></a>

## Build for Development

To build the project on your local machine, run:

    $ gulp serve

This runs some code quality checks, starts a web server at `http://localhost:9999` serving from the project root and generates the following files:

+ `./build/dev/prebid.js` - Full source code for dev and debug
+ `./build/dev/prebid.js.map` - Source map for dev and debug
+ `./build/dist/prebid.js` - Minified production code
+ `./prebid.js_<version>.zip` - Distributable zip archive

### Build Optimization

The standard build output contains all the available modules from within the `modules` folder.

You might want to exclude some/most of them from the final bundle.  To make sure the build only includes the modules you want, you can specify the modules to be included with the `--modules` CLI argument.

For example, when running the serve command: `gulp serve --modules=openxBidAdapter,rubiconBidAdapter,sovrnBidAdapter`

Building with just these adapters will result in a smaller bundle which should allow your pages to load faster.

**Build standalone prebid.js**

- Clone the repo, run `npm install`
- Then run the build:

        $ gulp build --modules=openxBidAdapter,rubiconBidAdapter,sovrnBidAdapter

Alternatively, a `.json` file can be specified that contains a list of modules you would like to include.

    $ gulp build --modules=modules.json

With `modules.json` containing the following
```json modules.json
[
  "openxBidAdapter",
  "rubiconBidAdapter",
  "sovrnBidAdapter"
]
```

**Build prebid.js using npm for bundling**

In case you'd like to explicitly show that your project uses `prebid.js` and want a reproducible build, consider adding it as an `npm` dependency.

- Add `prebid.js` as a `npm` dependency of your project: `npm install prebid.js`
- Run the `prebid.js` build under the `node_modules/prebid.js/` folder

        $ gulp build --modules=path/to/your/list-of-modules.json

Most likely your custom `prebid.js` will only change when there's:

- A change in your list of modules
- A new release of `prebid.js`

Having said that, you are probably safe to check your custom bundle into your project.  You can also generate it in your build process.

<a name="Run"></a>

## Test locally

To lint the code:

```bash
gulp lint
```

To run the unit tests:

```bash
gulp test
```

To generate and view the code coverage reports:

```bash
gulp test-coverage
gulp view-coverage
```

For end-to-end testing, edit the example file `./integrationExamples/gpt/pbjs_example_gpt.html`:

1. Change `{id}` values appropriately to set up ad units and bidders
2. Set the path to Prebid.js in your example file as shown below (see `pbs.src`).

For development:

```javascript
(function() {
    var d = document, pbs = d.createElement('script'), pro = d.location.protocol;
    pbs.type = 'text/javascript';
    pbs.src = ((pro === 'https:') ? 'https' : 'http') + './build/dev/prebid.js';
    var target = document.getElementsByTagName('head')[0];
    target.insertBefore(pbs, target.firstChild);
})();
```

For deployment:

```javascript
(function() {
    var d = document, pbs = d.createElement('script'), pro = d.location.protocol;
    pbs.type = 'text/javascript';
    pbs.src = ((pro === 'https:') ? 'https' : 'http') + './build/dist/prebid.js';
    var target = document.getElementsByTagName('head')[0];
    target.insertBefore(pbs, target.firstChild);
})();
```

Build and run the project locally with:

```bash
gulp serve
```

This runs `lint` and `test`, then starts a web server at `http://localhost:9999` serving from the project root.
Navigate to your example implementation to test, and if your `prebid.js` file is sourced from the `./build/dev`
directory you will have sourcemaps available in your browser's developer tools.

To run the example file, go to:

+ `http://localhost:9999/integrationExamples/gpt/pbjs_example_gpt.html`

As you make code changes, the bundles will be rebuilt and the page reloaded automatically.

<a name="Contribute"></a>

## Contribute

Many SSPs, bidders, and publishers have contributed to this project. [60+ Bidders](https://github.com/prebid/Prebid.js/tree/master/src/adapters) are supported by Prebid.js.

For guidelines, see [Contributing](./CONTRIBUTING.md).

Our PR review process can be found [here](https://github.com/prebid/Prebid.js/tree/master/pr_review.md).

### Add a Bidder Adapter

To add a bidder adapter module, see the instructions in [How to add a bidder adaptor](http://prebid.org/dev-docs/bidder-adaptor.html).

Please **do NOT load Prebid.js inside your adapter**. If you do this, we will reject or remove your adapter as appropriate.

### Code Quality

Code quality is defined by `.eslintrc` and errors are reported in the terminal.

If you are contributing code, you should [configure your editor](http://eslint.org/docs/user-guide/integrations#editors) with the provided `.eslintrc` settings.

### Unit Testing with Karma

        $ gulp test --watch --browsers=chrome

This will run tests and keep the Karma test browser open. If your `prebid.js` file is sourced from the `./build/dev` directory you will also have sourcemaps available when using your browser's developer tools.

+ To access the Karma debug page, go to `http://localhost:9876/debug.html`

+ For test results, see the console

+ To set breakpoints in source code, see the developer tools

Detailed code coverage reporting can be generated explicitly with

        $ gulp test --coverage

The results will be in

        ./build/coverage

*Note*: Starting in June 2016, all pull requests to Prebid.js need to include tests with greater than 80% code coverage before they can be merged.  For more information, see [#421](https://github.com/prebid/Prebid.js/issues/421).

For instructions on writing tests for Prebid.js, see [Testing Prebid.js](http://prebid.org/dev-docs/testing-prebid.html).

### Supported Browsers

Prebid.js is supported on IE10+ and modern browsers.

### Governance
Review our governance model [here](https://github.com/prebid/Prebid.js/tree/master/governance.md).


CHANGELOG
AOL Prebid 2.18.0
Updated to Prebid 1.17.0.


AOL Prebid 2.17.0
Updated to Prebid 1.16.0.


AOL Prebid 2.16.0
Updated to Prebid 1.15.0.


AOL Prebid 2.15.0
Updated to Prebid 1.14.0.


AOL Prebid 2.14.0
Updated to Prebid 1.13.0.


AOL Prebid 2.13.0
Fixed dropping pixels when userSyncOn="bidResponse".
Updated to Prebid 1.12.0.


AOL Prebid 2.12.0
Fixed dropping pixels when creatives are served into safe frames.
Improved public documentation.


AOL Prebid 2.11.0
----------------
Updated to Prebid 1.11.0
Added GDPR support fo AOL analytics adapter.
Added passing GDPR flag when consent string is not present.
Fixed wrong bid status issue for empty bids in AOL analytics adapter.


AOL Prebid 2.10.0
----------------
Updated to Prebid 1.10.0


AOL Prebid 2.9.0
----------------
Updated to Prebid 1.9.0
Added GDPR support for AOL adapter.


AOL Prebid 2.8.0
----------------
Updated to Prebid 1.8.0


AOL Prebid 2.7.0
----------------
Updated to Prebid 1.7.0


AOL Prebid 2.6.0
----------------
Updated to Prebid 1.6.0


AOL Prebid 2.5.0
----------------
Updated to Prebid 1.5.0


AOL Prebid 2.4.0
----------------
Updated to Prebid 1.4.0


AOL Prebid 2.3.0
----------------
Updated to Prebid 1.3.1


AOL Prebid 2.2.0
----------------
Updated to Prebid 1.2.0


AOL Prebid 2.1.0
----------------
Updated to Prebid 1.1.1


AOL Prebid 2.0.0
----------------
Updated to Prebid 1.0.0


AOL Prebid 1.36.0
----------------
Fixed issue with dropping pixels in nested iframes.


AOL Prebid 1.35.0
----------------
Fixed failing AOL analytics adapter when passing "adUnits" into "requestBids".


AOL Prebid 1.34.0
----------------
Updated to Prebid 0.34.1.
Changed ttl for aol adapter endpoints.


AOL Prebid 1.33.0
----------------
Updated to Prebid 0.33.0


AOL Prebid 1.32.0
----------------
Updated to Prebid 0.32.0


AOL Prebid 1.31.0
----------------
Updated to Prebid 0.31.0


AOL Prebid 1.30.0
----------------
Updated to Prebid 0.30.1


AOL Prebid 1.29.0
----------------
Updated to Prebid 0.29.0
Updated AOL adapter to be compatible with Prebid 1.0.


AOL Prebid 1.28.0
----------------
Updated to Prebid 0.28.0


AOL Prebid 1.27.0
----------------
Updated Komoona adapter.
Updated to Prebid 0.27.1


AOL Prebid 1.26.0
----------------
Added restrictions for adIdExtension field.
Added partners ids for s2s support.


AOL Prebid 1.25.0
----------------
Updated to Prebid 0.26.1


AOL Prebid 1.24.0
----------------
Updated to Prebid 0.25.0


AOL Prebid 1.23.0
----------------
Added passing key values feature.
Updated to Prebid 0.24.1


AOL Prebid 1.22.0
----------------
Added aliases for AOL adapter.


AOL Prebid 1.21.0
----------------
Updated to Prebid 0.23.1


AOL Prebid 1.20.0
----------------
Updated to Prebid 0.22.2


AOL Prebid 1.19.0
----------------
Updated to Prebid 0.21.0


AOL Prebid 1.18.0
----------------
Updated to Prebid 0.20.0


AOL Prebid 1.17.0
----------------
Added functionality for rendering pixels once.


AOL Prebid 1.16.0
----------------
Added Audience Network adapter.
Added creative key field in bid response for Sharethrough adapter.


AOL Prebid 1.15.0
----------------
Nexage API implemented.


AOL Prebid 1.14.0
----------------
Updated to Prebid 0.19.0


AOL Prebid 1.13.0
----------------
Updated to Prebid 0.18.0


AOL Prebid 1.12.0
----------------
Updated to Prebid 0.17.0


AOL Prebid 1.11.0
----------------
Added functionality of cookie syncing upon bid response.


AOL Prebid 1.10.0
----------------
Updated to Prebid 0.16.0


AOL Prebid 1.9.1
----------------
Bump to patch version due to issue on build.


AOL Prebid 1.9.0
----------------
Add AdID field extension for reporting.


AOL Prebid 1.8.0
----------------
Updated to Prebid 0.15.2


AOL Prebid 1.7.0
----------------
Updated to Prebid 0.14.0


AOL Prebid 1.6.0
----------------
Support for reporting bid identifiers.
BUGFIX: indexExchange has now the correct bidder code.


AOL Prebid 1.5.1
----------------
BUGFIX: Fixed accidental build error associated with window=window.


AOL Prebid 1.5.0
----------------
Reporting support for Deals.
BUGFIX: Source version now has no errors associated with window=window.


AOL Prebid 1.4.2
----------------
BUGFIX: Fixed reporting refreshed ad units in AOL analytics.


AOL Prebid 1.4.1
----------------
Refactoring of AOL analytics.
Fixed reporting of pageId parameter in AOL analytics.
BUGFIX: Fixed retrieving 3rd party pixels from bid response in AOL adapter.


AOL Prebid 1.4.0
----------------
Updated to Prebid 0.13.1


AOL Prebid 1.3.1
----------------
Fixed reporting of timed out bids in AOL analytics.


AOL Prebid 1.3.0
----------------
Updated to Prebid 0.13.0.


AOL Prebid 1.2.0
----------------
Updated to Prebid 0.12.0.
Fixed sorting of per-adapter timeouts.
Fixed ES6 features which require a polyfill.


AOL Prebid 1.1.0
----------------
AOL adapter is optimized and heavily reduced in size.
Small code optimizations in AOL analytics adapter.
AOL analytics adapter now includes currency in the hbevent requests.


AOL Prebid 1.0.7
----------------
BUGFIX: AOL analytics adapter reports timed out bids.


AOL Prebid 1.0.6
----------------
BUGFIX: Ensure that auctionId never exceeds max 64 bit value. Limit length to 18 digits.


AOL Prebid 1.0.5
----------------
BUGFIX: Ensure that an event is logged in the case when there are no bids from the DSPs.
Populate the bid object with the deal ID for private deals.


AOL Prebid 1.0.4
----------------
BUGFIX: Fixed default server for AOL adapter & analytics.
BUGFIX: Fixed region option to be picked up by AOL analytics.
BUGFIX: Separate server host names used for AOL adapter & analytics.
BUGFIX: Consolidated default parameter values for invalid bids in AOL analytics.


AOL Prebid 1.0.3
----------------
BUGFIX: Fixed empty hbbid parameter in AOL analytics in case of invalid bid response.


AOL Prebid 1.0.2
----------------
BUGFIX: Fixed error in AOL adapter when processing cpm from bid response.
BUGFIX: Support for undefined options in AOL analytics.


AOL Prebid 1.0.1

gulpfile.js
'use strict';

var _ = require('lodash');
var argv = require('yargs').argv;
var gulp = require('gulp');
var gutil = require('gulp-util');
var connect = require('gulp-connect');
var path = require('path');
var webpack = require('webpack');
var webpackStream = require('webpack-stream');
var uglify = require('gulp-uglify');
var clean = require('gulp-clean');
var KarmaServer = require('karma').Server;
var karmaConfMaker = require('./karma.conf.maker');
var opens = require('open');
var webpackConfig = require('./webpack.conf');
var helpers = require('./gulpHelpers');
var del = require('del');
var gulpDocumentation = require('gulp-documentation');
var concat = require('gulp-concat');
var header = require('gulp-header');
var footer = require('gulp-footer');
var replace = require('gulp-replace');
var shell = require('gulp-shell');
var optimizejs = require('gulp-optimize-js');
var eslint = require('gulp-eslint');
var gulpif = require('gulp-if');
var sourcemaps = require('gulp-sourcemaps');
var through = require('through2');
var fs = require('fs');
var jsEscape = require('gulp-js-escape');
const karmaConfigProvider = require('./karmaConfigProvider');

var prebid = require('./package.json');
var banner = '/* <%= prebid.name %> v<%= prebid.version %>*/\n';
var analyticsDirectory = '../analytics';
var port = 9999;

// Tasks
gulp.task('default', ['webpack']);

gulp.task('serve', ['lint', 'build-bundle-dev', 'watch', 'test']);

gulp.task('serve-nw', ['lint', 'watch', 'e2etest']);

gulp.task('run-tests', ['lint', 'test-coverage']);

gulp.task('build', ['build-bundle-prod']);

gulp.task('clean', function () {
  return gulp.src(['build'], {
      read: false
    })
    .pipe(clean());
});

function gulpBundle(dev) {
  return bundle(dev).pipe(gulp.dest('build/' + (dev ? 'dev' : 'dist')));
}

function nodeBundle(modules) {
  return new Promise((resolve, reject) => {
    bundle(false, modules)
      .on('error', (err) => {
        reject(err);
      })
      .pipe(through.obj(function(file, enc, done) {
        resolve(file.contents.toString(enc));
        done();
      }));
  });
}

// these modules must be explicitly listed in --modules to be included in the build, won't be part of "all" modules
var explicitModules = [
  'pre1api'
];

function bundle(dev, moduleArr) {
  var modules = moduleArr || helpers.getArgModules(),
      allModules = helpers.getModuleNames(modules);

  if(modules.length === 0) {
    modules = allModules.filter(module => !explicitModules.includes(module));
  } else {
    var diff = _.difference(modules, allModules);
    if(diff.length !== 0) {
      throw new gutil.PluginError({
        plugin: 'bundle',
        message: 'invalid modules: ' + diff.join(', ')
      });
    }
  }

  var entries = [helpers.getBuiltPrebidCoreFile(dev)].concat(helpers.getBuiltModules(dev, modules));

  var outputFileName = argv.bundleName ? argv.bundleName : 'prebid.js';

  // change output filename if argument --tag given
  if (argv.tag && argv.tag.length) {
    outputFileName = outputFileName.replace(/\.js$/, `.${argv.tag}.js`);
  }

  gutil.log('Concatenating files:\n', entries);
  gutil.log('Appending ' + prebid.globalVarName + '.processQueue();');
  gutil.log('Generating bundle:', outputFileName);

  return gulp.src(
      entries
    )
    .pipe(gulpif(dev, sourcemaps.init({loadMaps: true})))
    .pipe(concat(outputFileName))
    .pipe(gulpif(!argv.manualEnable, footer('\n<%= global %>.processQueue();', {
        global: prebid.globalVarName
      }
    )))
    .pipe(gulpif(dev, sourcemaps.write('.')));
}

// Workaround for incompatibility between Karma & gulp callbacks.
// See https://github.com/karma-runner/gulp-karma/issues/18 for some related discussion.
function newKarmaCallback(done) {
  return function (exitCode) {
    if (exitCode) {
      done(new Error('Karma tests failed with exit code ' + exitCode));
    } else {
      if (argv.browserstack) {
        process.exit(0);
      } else {
        done();
      }
    }
  }
}

gulp.task('build-bundle-dev', ['devpack'], gulpBundle.bind(null, true));
gulp.task('build-bundle-prod', ['webpack'], gulpBundle.bind(null, false));
gulp.task('bundle', gulpBundle.bind(null, false)); // used for just concatenating pre-built files with no build step

gulp.task('bundle-to-stdout', function() {
  nodeBundle().then(file => console.log(file));
});

gulp.task('devpack', ['clean'], function () {
  var cloned = _.cloneDeep(webpackConfig);
  cloned.devtool = 'source-map';
  var externalModules = helpers.getArgModules();

  const analyticsSources = helpers.getAnalyticsSources(analyticsDirectory);
  const moduleSources = helpers.getModulePaths(externalModules);

  return gulp.src([].concat(moduleSources, analyticsSources, 'src/prebid.js'))
    .pipe(helpers.nameModules(externalModules))
    .pipe(webpackStream(cloned, webpack))
    .pipe(replace('$prebid.version$', prebid.version))
    .pipe(gulp.dest('build/dev'))
    .pipe(connect.reload());
});

gulp.task('webpack', ['clean'], function () {
  var cloned = _.cloneDeep(webpackConfig);

  delete cloned.devtool;

  var externalModules = helpers.getArgModules();

  const analyticsSources = helpers.getAnalyticsSources(analyticsDirectory);
  const moduleSources = helpers.getModulePaths(externalModules);

  return gulp.src([].concat(moduleSources, analyticsSources, 'src/prebid.js'))
    .pipe(helpers.nameModules(externalModules))
    .pipe(webpackStream(cloned, webpack))
    .pipe(replace('$prebid.version$', prebid.version))
    .pipe(uglify())
    .pipe(gulpif(file => file.basename === 'prebid-core.js', header(banner, { prebid: prebid })))
    .pipe(optimizejs())
    .pipe(gulp.dest('build/dist'))
    .pipe(connect.reload());
});

// Run the unit tests.
//
// By default, this runs in headless chrome.
//
// If --watch is given, the task will re-run unit tests whenever the source code changes
// If --file "<path-to-test-file>" is given, the task will only run tests in the specified file.
// If --browserstack is given, it will run the full suite of currently supported browsers.
// If --browsers is given, browsers can be chosen explicitly. e.g. --browsers=chrome,firefox,ie9
gulp.task('test', ['clean'], function (done) {
  var karmaConf = karmaConfMaker(false, argv.browserstack, argv.watch, argv.file);

  var browserOverride = helpers.parseBrowserArgs(argv).map(helpers.toCapitalCase);
  if (browserOverride.length > 0) {
    karmaConf.browsers = browserOverride;
  }

  new KarmaServer(karmaConf, newKarmaCallback(done)).start();
});

// If --file "<path-to-test-file>" is given, the task will only run tests in the specified file.
gulp.task('test-coverage', ['clean'], function(done) {
  new KarmaServer(karmaConfMaker(true, false, false, argv.file), newKarmaCallback(done)).start();
});

// View the code coverage report in the browser.
gulp.task('view-coverage', function (done) {
  var coveragePort = 1999;

  connect.server({
    port: coveragePort,
    root: 'build/coverage/karma_html',
    livereload: false
  });
  opens('http://localhost:' + coveragePort);
  done();
});

gulp.task('coveralls', ['test-coverage'], function() { // 2nd arg is a dependency: 'test' must be finished
  // first send results of istanbul's test coverage to coveralls.io.
  return gulp.src('gulpfile.js', { read: false }) // You have to give it a file, but you don't
  // have to read it.
    .pipe(shell('cat build/coverage/lcov.info | node_modules/coveralls/bin/coveralls.js'));
});

// Watch Task with Live Reload
gulp.task('watch', function () {
  gulp.watch([
    'src/**/*.js',
    'modules/**/*.js',
    'test/spec/**/*.js',
    '!test/spec/loaders/**/*.js'
  ], ['lint', 'build-bundle-dev', 'test']);
  gulp.watch([
    'loaders/**/*.js',
    'test/spec/loaders/**/*.js'
  ], ['lint']);
  connect.server({
    https: argv.https,
    port: port,
    root: './',
    livereload: true
  });
});

gulp.task('lint', () => {
  return gulp.src(['src/**/*.js', 'modules/**/*.js', 'test/**/*.js'])
    .pipe(eslint())
    .pipe(eslint.format('stylish'))
    .pipe(eslint.failAfterError());
});

gulp.task('clean-docs', function () {
  del(['docs']);
});

gulp.task('docs', ['clean-docs'], function () {
  return gulp.src('src/prebid.js')
    .pipe(gulpDocumentation('md'))
    .on('error', function (err) {
      gutil.log('`gulp-documentation` failed:', err.message);
    })
    .pipe(gulp.dest('docs'));
});

gulp.task('e2etest', ['devpack', 'webpack'], function() {
  var cmdQueue = [];
  if(argv.browserstack) {
    var browsers = require('./browsers.json');
    delete browsers['bs_ie_9_windows_7'];

    var cmdStr = ' --config nightwatch.conf.js';
    if (argv.group) {
      cmdStr = cmdStr + ' --group ' + argv.group;
    }
    cmdStr = cmdStr + ' --reporter ./test/spec/e2e/custom-reporter/pbjs-html-reporter.js';

    var startWith = 'bs';

    Object.keys(browsers).filter(function(v){
      return v.substring(0, startWith.length) === startWith && browsers[v].browser !== 'iphone';
    }).map(function(v,i,arr) {
      var newArr = (i%2 === 0) ? arr.slice(i,i+2) : null;
      if(newArr) {
        var cmd = 'nightwatch --env ' + newArr.join(',') + cmdStr;
        cmdQueue.push(cmd);
      }
    });
  }

  return gulp.src('')
    .pipe(shell(cmdQueue.join(';')));
});

gulp.task('e2etest-report', function() {
  var reportPort = 9010;
  var targetDestinationDir = './e2etest-report';
  helpers.createEnd2EndTestReport(targetDestinationDir);
  connect.server({
    port: reportPort,
    root: './',
    livereload: true
  });

  setTimeout(function() {
    opens('http://localhost:' + reportPort + '/' + targetDestinationDir.slice(2) + '/results.html');
  }, 5000);
});

// This task creates postbid.js. Postbid setup is different from prebid.js
// More info can be found here http://prebid.org/overview/what-is-post-bid.html
gulp.task('build-postbid', ['escape-postbid-config'], function() {
  var fileContent = fs.readFileSync('./build/postbid/postbid-config.js', 'utf8');

  return gulp.src('./integrationExamples/postbid/oas/postbid.js')
    .pipe(replace('\[%%postbid%%\]', fileContent))
    .pipe(gulp.dest('build/postbid/'));
});

// Dependant task for building postbid. It escapes postbid-config file.
gulp.task('escape-postbid-config', function() {
  gulp.src('./integrationExamples/postbid/oas/postbid-config.js')
    .pipe(jsEscape())
    .pipe(gulp.dest('build/postbid/'));
});

gulp.task('build-aol-bundle', ['build-bundle-dev'], () => {
  return gulp.src('build/dev/prebid.js')
    .pipe(replace(/(\/\*!(ANALYTICS\s)?ADAPTER BEGIN \w+\*\/)/g, '$1window=window;'))
    .pipe(replace(/(\/\*!(ANALYTICS\s)?ADAPTER END \w+\*\/)/g, '$1window=window;'))
    .pipe(uglify({
      output: {
        comments: /^!/
      }
    }))
    .pipe(optimizejs())
    .pipe(replace(/(\/\*!(ANALYTICS\s)?ADAPTER BEGIN \w+\*\/)\s*window=window(;|,)?/g, '$1'))
    .pipe(replace(/(,)?(\/\*!(ANALYTICS\s)?ADAPTER END \w+\*\/)\s*window=window(;|,)?/g, ';$2'))
    .pipe(header(banner, { prebid: prebid }))
    .pipe(gulp.dest('build/dist'));
});

gulp.task('test-unit', ['clean'], function (done) {
  new KarmaServer(karmaConfigProvider.generateDefaultConfig(), newKarmaCallback(done)).start();
});

gulp.task('test-unit-cloud', ['clean'], function (done) {
  if (!process.env.SAUCE_USERNAME || !process.env.SAUCE_ACCESS_KEY) {
    done('Make sure the SAUCE_USERNAME and SAUCE_ACCESS_KEY environment variables are set.');
  } else {
    new KarmaServer(karmaConfigProvider.generateSaucelabsConfig(), newKarmaCallback(done)).start();
  }
});

module.exports = nodeBundle;
